libcloud/storage/base.py

```diff
@@ -622,200 +624,62 @@ def _upload_object(self, object_name, content_type, upload_func,
                     # Fallback to a content-type
                     content_type = DEFAULT_CONTENT_TYPE
 
-        file_size = None
-
-        if iterator:
-            if self.supports_chunked_encoding:
-                headers['Transfer-Encoding'] = 'chunked'
-                upload_func_kwargs['chunked'] = True
-            else:
-                # Chunked transfer encoding is not supported. Need to buffer
-                # all the data in memory so we can determine file size.
-                iterator = libcloud.utils.files.read_in_chunks(
-                    iterator=iterator)
-                data = libcloud.utils.files.exhaust_iterator(iterator=iterator)
-
-                file_size = len(data)
-                upload_func_kwargs['data'] = data
-        else:
-            file_size = os.path.getsize(file_path)
-            upload_func_kwargs['chunked'] = False
-
-        if file_size is not None and 'Content-Length' not in headers:
-            headers['Content-Length'] = file_size
-
         headers['Content-Type'] = content_type
-        response = self.connection.request(request_path,
-                                           method=request_method, data=None,
-                                           headers=headers, raw=True)
-
-        upload_func_kwargs['response'] = response
-        success, data_hash, bytes_transferred = upload_func(
-            **upload_func_kwargs)
-
-        if not success:
+        if stream:
+            response = self.connection.request(
+                request_path,
+                method=request_method, data=stream,
+                headers=headers, raw=True)
+            stream_hash, stream_length = self._hash_buffered_stream(
+                stream,
```

Comment:
So, if stream is a generator (it has `next`), does consuming it twice actually work? I would expect it to be empty the second time around, as generators generally do not generate the same content twice. This contrary to the file code below, which is clear.